KOGITO-8445: SWF Diagram Viewer - Orthogonal lines

handreyrc opened the pull request:
Hey @romartin;
Could you please review this PR?

 **JIRA**: [KOGITO-8445](https://issues.redhat.com/browse/KOGITO-8445)

Thanks!

romartin reviewed (commented):
Code looks good! Thanks @handreyrc 

I will proceed with the UI testing as well.

Please @LuboTerifaj this is quite urgent, deadline is end of this week. Could you please start the review asap? Thanks in advance!

romartin reviewed (approved):
Hey @handreyrc 

I just tested the UI with the examples from kubesmarts and also some other examples. In general they're looking much better, thanks! 

There are a few UI improvments, as expected, but in my opinion those can be postponed for further improvements. For instance, in some edge case, the "green" connector that goes back to a previous state is a bit weird, but the workflow anyway can be properly read and there exist no overlappings, as before, so great! 

![Screenshot from 2023-02-13 21-49-01](https://user-images.githubusercontent.com/4602417/218573071-b62088c1-0aa4-4d01-a747-86a58fd591ed.png)

Also another issue I noticed, althought not related to this PR, is that the label text sometimes overlaps itself, when multiple lines, as you can also see on the screenshot above.

Anyway IMO it is much better than before, and it is ready to go!! Thanks for this work!

@LuboTerifaj on your hands now, please consider giving us early feedback , so if any issue, we should have still time for fixing.

Thanks guys!

LuboTerifaj reviewed (changes requested):
Hello @handreyrc , @romartin ,

### Found issues:

The graph is not centered when there are more levels with more states.
- See [book-lending.sw.json](https://github.com/LuboTerifaj/specification/blob/examples-in-separate-files/examples/resources/book-lending.sw.json), [monitor-job.sw.json](https://github.com/LuboTerifaj/specification/blob/examples-in-separate-files/examples/resources/monitor-job.sw.json) from specification

The outgoing connection to the node that is in the same or higher level is pinned to the bottom magnet instead of west/east magnet.
- The actual behaviour conforms with the jira description, but it would make much more sense to use east/west magnets.
- See [monitor-job.sw.json](https://github.com/LuboTerifaj/specification/blob/examples-in-separate-files/examples/resources/monitor-job.sw.json) from specification

If the connection is going from source to target directly to level above, it is going from west magnet. It might look better if it went directly from the top magnet.
- See [car-vitals-checks-parent.sw.json](https://github.com/LuboTerifaj/specification/blob/examples-in-separate-files/examples/resources/car-vitals-checks-parent.sw.json), [continuing-as-a-new-execution.sw.json](https://github.com/LuboTerifaj/specification/blob/examples-in-separate-files/examples/resources/continuing-as-a-new-execution.sw.json), [filling-a-glass-of-water.sw.json](https://github.com/LuboTerifaj/specification/blob/examples-in-separate-files/examples/resources/filling-a-glass-of-water.sw.json) from specification

There is an **error in diagram**.
- Error Message: "_Transition from $STATE_NAME has invalid target._"
- The end node is not shown in the diagram.
- It happens if there is `"end": true` specified in `dataConditions`, `defualtconditions`, `eventConditions`, or `onErrors`
- See [book-lending.sw.json](https://github.com/LuboTerifaj/specification/blob/examples-in-separate-files/examples/resources/book-lending.sw.json), [car-vitals-checks-parent.sw.json](https://github.com/LuboTerifaj/specification/blob/examples-in-separate-files/examples/resources/car-vitals-checks-parent.sw.json), [filling-a-glass-of-water.sw.json](https://github.com/LuboTerifaj/specification/blob/examples-in-separate-files/examples/resources/filling-a-glass-of-water.sw.json), [new-patient-onboarding.sw.json](https://github.com/LuboTerifaj/specification/blob/examples-in-separate-files/examples/resources/new-patient-onboarding.sw.json) from specification
- See [currency-exchange-workflow.sw.json](https://github.com/kiegroup/kogito-examples/tree/stable/serverless-workflow-examples/serverless-workflow-oauth2-orchestration-quarkus/currency-exchange-workflow/src/main/resources/currency-exchange-workflow.sw.json), [resources/handle-frontend.sw.json](https://github.com/kiegroup/kogito-examples/tree/stable/serverless-workflow-examples/serverless-workflow-github-showcase/pr-checker-workflow/src/main/resources/handle-frontend.sw.json), [handle-backend.sw.json](https://github.com/kiegroup/kogito-examples/tree/stable/serverless-workflow-examples/serverless-workflow-github-showcase/pr-checker-workflow/src/main/resources/handle-backend.sw.json), [callback.sw.json](https://github.com/kiegroup/kogito-examples/tree/stable/serverless-workflow-examples/serverless-workflow-callback-quarkus/src/main/resources/callback.sw.json) from kogito-examples

There is an **error in diagram**.
- Error Message: "_Transition from $STATE_NAME has invalid target._"
- The end node is not shown in the diagram.
- It happens if there is `"terminate": true` used in workflow.
- See [loanbroker.sw.json](https://github.com/kiegroup/kogito-examples/tree/stable/serverless-workflow-examples/serverless-workflow-loanbroker-showcase/loanbroker-flow/src/main/resources/loanbroker.sw.json) from kogito-examples

The connection is going through states.
- See [monitor-job.sw.json](https://github.com/LuboTerifaj/specification/blob/examples-in-separate-files/examples/resources/monitor-job.sw.json), [order-saga-error-handling.sw.json](https://github.com/kiegroup/kogito-examples/tree/stable/serverless-workflow-examples/serverless-workflow-saga-quarkus/src/main/resources/order-saga-error-handling.sw.json) from specification

The connection crosses other connections
- See [monitor-job.sw.json](https://github.com/LuboTerifaj/specification/blob/examples-in-separate-files/examples/resources/monitor-job.sw.json) from specification

The connections are going through states/nodes in some scenarios if there are multiple states/nodes in the same level, depending on node distribution.
There is also an issue with connections when the states don't have incoming connections, so the nodes are shown next to the Start node. As a result, the connections are going through the same level states.
- Consider the following example:
<details><summary>example.sw.json</summary>

```
{
   "id": "helloworld",
   "version": "1.0",
   "specVersion": "0.8",
   "name": "Hello World Workflow",
   "description": "Inject Hello World",
   "start": "State1",
   "states":[
     {
        "name":"State1",
        "type":"inject",
        "data": {
           "result": "Hello World!"
        },
        "transition":"State2"
     },
     {
      "name":"State without both transitions",
      "type":"inject",
      "data": {
         "result": "Hello World!"
      }        
   },
     {
         "name":"State without transition",
         "type":"inject",
         "data": {
            "result": "Hello World!"
         },
         "transition": "State without transition 2"        
      },
      {
         "name":"State without transition 3",
         "type":"inject",
         "data": {
            "result": "Hello World!"
         },
         "end":true  
      },
      {
         "name":"State without transition 1",
         "type":"inject",
         "data": {
            "result": "Hello World!"
         },
         "transition": "State without transition 2"
      },
      {
         "name":"State without transition 2",
         "type":"inject",
         "data": {
            "result": "Hello World!"
         },
         "transition": "State3"  
      },
      {
         "name":"State2",
         "type":"inject",
         "data": {
            "result": "Hello World!"
         },
         "transition": "State3"  
      },
      {
         "name":"State3",
         "type":"inject",
         "data": {
            "result": "Hello World!"
         },
         "end":true  
      },      
      {
         "name":"State without transition 4",
         "type":"inject",
         "data": {
            "result": "Hello World!"
         },
         "transition":"State1"  
      }
   ]
   }

```
</details>

### Comments

End nodes at the bottom level are not connected with a single simple line. The polyline is used instead.
- See compensation.sw.json from https://start.kubesmarts.org/#/import?url=https://start.kubesmarts.org/samples/compensation/compensation.sw.json&renameWorkspace=Compensation
- I understand this may be wanted as the end nodes are centered.

The connection label is shown using two rows. In addition there is redundant blue box area around the text.
- See [monitor-job.sw.json](https://github.com/LuboTerifaj/specification/blob/examples-in-separate-files/examples/resources/monitor-job.sw.json), [reusing-function-and-event-definitions.sw.json](https://github.com/LuboTerifaj/specification/blob/examples-in-separate-files/examples/resources/reusing-function-and-event-definitions.sw.json)
- See [handle-frontend.sw.json](https://github.com/kiegroup/kogito-examples/tree/stable/serverless-workflow-examples/serverless-workflow-github-showcase/pr-checker-workflow/src/main/resources/handle-frontend.sw.json), [handle-backend.sw.json](https://github.com/kiegroup/kogito-examples/tree/stable/serverless-workflow-examples/serverless-workflow-github-showcase/pr-checker-workflow/src/main/resources/handle-backend.sw.json)
- It may look good showing it in two rows, but the redundant box should be removed and the text should be split "symmetrically" instead of just letting the one character in the new row. In addition, it should be split only in case there is not enough space (space from UX point of view).

Overall experience is much better than before.

Thank you!

romartin commented:
THanks @LuboTerifaj for great review and feedback! I will let @handreyrc to asnwer your comment above, he's going to fix the issue and report others in a ticket for improvements. Also @handreyrc added new integration tests about control points, which is great, thanks @handreyrc !

handreyrc commented:
Hey @LuboTerifaj and @romartin,

I went through all the points you raised. I agree we still have improvents to be done.
Anyway it's a lot of work to be done in a single PR so I created a new JIRA for them.
WDYT?
Please feel free to add more points you find relevant to be addressed.
**JIRA**: [KOGITO-86810](https://issues.redhat.com/browse/KOGITO-86810)

About the errors presented in the diagram.
One can route to End state from state's declaration and from transition's declaration as well. Only the first case was covered before. I implemented the End node creation from transitions so it is fixed.

BTW I also added IT tests for the control points created by the orthogonal lines.

We are ready for a sencond round!

Thank you guys for looking into it!

LuboTerifaj commented:
Hello @handreyrc ,

I am not sure which issues did you fix, and which are you suggesting to implement in a separate jira.

Could you please clarify it to me?

Please don't report jiras with the issues the reviewer found until we agree on fixing them in separate PR.

Thanks!

handreyrc commented:
Hey @LuboTerifaj,
I fixed only the two items you mentioned there are errors in the diagram. 

There is an error in diagram.

Error Message: "Transition from $STATE_NAME has invalid target."
The end node is not shown in the diagram.
It happens if there is "end": true specified in dataConditions, defualtconditions, eventConditions, or onErrors
See [book-lending.sw.json](https://github.com/LuboTerifaj/specification/blob/examples-in-separate-files/examples/resources/book-lending.sw.json), [car-vitals-checks-parent.sw.json](https://github.com/LuboTerifaj/specification/blob/examples-in-separate-files/examples/resources/car-vitals-checks-parent.sw.json), [filling-a-glass-of-water.sw.json](https://github.com/LuboTerifaj/specification/blob/examples-in-separate-files/examples/resources/filling-a-glass-of-water.sw.json), [new-patient-onboarding.sw.json](https://github.com/LuboTerifaj/specification/blob/examples-in-separate-files/examples/resources/new-patient-onboarding.sw.json) from specification
See [currency-exchange-workflow.sw.json](https://github.com/kiegroup/kogito-examples/tree/stable/serverless-workflow-examples/serverless-workflow-oauth2-orchestration-quarkus/currency-exchange-workflow/src/main/resources/currency-exchange-workflow.sw.json), [resources/handle-frontend.sw.json](https://github.com/kiegroup/kogito-examples/tree/stable/serverless-workflow-examples/serverless-workflow-github-showcase/pr-checker-workflow/src/main/resources/handle-frontend.sw.json), [handle-backend.sw.json](https://github.com/kiegroup/kogito-examples/tree/stable/serverless-workflow-examples/serverless-workflow-github-showcase/pr-checker-workflow/src/main/resources/handle-backend.sw.json), [callback.sw.json](https://github.com/kiegroup/kogito-examples/tree/stable/serverless-workflow-examples/serverless-workflow-callback-quarkus/src/main/resources/callback.sw.json) from kogito-examples
There is an error in diagram.

Error Message: "Transition from $STATE_NAME has invalid target."
The end node is not shown in the diagram.
It happens if there is "terminate": true used in workflow.
See [loanbroker.sw.json](https://github.com/kiegroup/kogito-examples/tree/stable/serverless-workflow-examples/serverless-workflow-loanbroker-showcase/loanbroker-flow/src/main/resources/loanbroker.sw.json) from kogito-examples

Thanks!

LuboTerifaj reviewed (approved):
Hello @handreyrc ,

Please have a look at my inline comments. Nothing significant, but I would like to know your opinion on that. It doesn't need to be done in this PR.

The most critical issues have been fixed. I reported the unfixed issues I found here:

- [KOGITO-8683](https://issues.redhat.com/browse/KOGITO-8683) _Orthogonal lines improvements_
- [KOGITO-8684](https://issues.redhat.com/browse/KOGITO-8684) _Connection label on hover is not shown properly_

Thank you!

romartin commented:
Great, thanks @LuboTerifaj for your great work and collaboration here!

handreyrc commented on packages/serverless-workflow-diagram-editor/sw-editor/sw-editor-kogito-app/src/test/java/org/kie/workbench/common/stunner/sw/client/selenium/SWEditorSeleniumIT.java:
```diff
@@ -182,6 +184,164 @@ public void testProcessTransactionsExample() throws Exception {
+
+        // First connector
+        List<Object> points = jsHelper.getConnectorPoints(nodeIds.get(1), 0, 0);
+        assertThat(points.size()).isEqualTo(10);
```
Hey @LuboTerifaj,

These are the parameters to get to an specific connection associated to a magnet index of a node we are interested in.
It's bit tricky but it is how lienzo internals work. You have to isolate a magnet and then get to the connections associated to it.

To get to those hardcoded values there is no other way than load the diagram and from the console use the js canvas API to isolate the nodes indexes and connections you want to test.

Please take a look at the method "getConnectorPoints" implemented in the JsCanvasHelper class.
e.g.
getConnectorPoints([node uuid], [magnet 0-center, 1-top, 2-left, 3-bottom, 4-right], [index of the connection])

LuboTerifaj replied:
I meant the values like `10`, `558.5348837209302D`, `319L`, etc.. It is not easily readable why those values are expected in the assertions. I would move them to constants, or at least document, why those values are expected. The assertions used are not readable well with these values.

handreyrc replied:
I see your point. 
I definitely could have added some explanatory comments about those values.
It's not that clear they are X, Y coordinates for the points we are recovering from the connector.
Thanks!

handreyrc commented on packages/serverless-workflow-diagram-editor/sw-editor/sw-editor-api/src/main/java/org/kie/workbench/common/stunner/sw/autolayout/AutoLayout.java:
```diff
@@ -87,17 +90,246 @@ public static Promise<Node> applyLayout(final Graph graph,
     }
 
+    @SuppressWarnings("all")
+    static void moveEndNodesX(Graph graph, final double x) {
```
Hey @LuboTerifaj,

These methods are being called internally in a deep chain of promisses which leads to never ending list of cast issues if somehow you try to mock them. I wouldn't say it is impossible to mock them, but it is a lot more work than the actual implementation once we are interacting with a mocked graph instance.
